### PR TITLE
Dynamic API

### PR DESCRIPTION
**Pull request by cmelchior**
This PR is a prerequisite for the new Migration API. It adds support for manipulating RealmObjects/RealmList using a Dynamic API instead of the current type safe one.

Implementation details:
DynamicRealmList currently only extends AbstractList instead of RealmList. This is to avoid also having to make a dynamic query API as well, which seemed a bit overkill for migration purposes. For a full blown dynamic API it will be needed, but for the purpose of Migrations I think it can be left out.

```
public interface DynamicRealmObjectInterface {
    boolean getBoolean(String fieldName);
    int getInt(String fieldName);
    short getShort(String fieldName);
    long getLong(String fieldName);
    byte getByte(String fieldName);
    float getFloat(String fieldName);
    double getDouble(String fieldName);
    byte[] getBlob(String fieldName);
    String getString(String fieldName);
    Date getDate(String fieldName);
    DynamicRealmObject getObject(String fieldName);
    DynamicRealmList getList(String fieldName);
    void setBoolean(String fieldName, boolean value);
    void setShort(String fieldName, short value);
    void setInt(String fieldName, int value);
    void setLong(String fieldName, long value);
    void setByte(String fieldName, byte value);
    void setFloat(String fieldName, float value);
    void setDouble(String fieldName, double value);
    void setString(String fieldName, String value);
    void setBlob(String fieldName, byte[] value);
    void setDate(String fieldName, Date value);
    void setObject(String fieldName, DynamicRealmObject value);
    void setList(String fieldName, DynamicRealmList list);
    boolean isNull(String fieldName);
    boolean hasField(String fieldName);
    String[] getFieldNames();
    void removeFromRealm();
}
```
 How to support the difference between null fields and not-nullable fields? Solution: add ```isNull``` for getters. Override setters: ```setInt(int)``` and ```setInt(Integer)```.

### CODE REVIEW

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
+package io.realm;/*
+ * Copyright 2014 Realm Inc.
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: 2015

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
+package io.realm;/*
+ * Copyright 2014 Realm Inc.
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: 2015

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2014 Realm Inc.
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: 2015

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        linkView.set(location, object.row.getIndex());
+        return object;
+    }
+
+    @Override
+    public int size() {
+        return ((Long)linkView.size()).intValue();
+    }
+
+    private void checkIsValidObject(DynamicRealmObject object) {
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: I would rename this to `isObjectValid()`

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            throw new IllegalArgumentException("Cannot add a object belonging already in another Realm");
+        }
+        if (!linkView.getTable().hasSameSchema(object.row.getTable())) {
+            String expectedClass = linkView.getTable().getName();
+            String objectClassName = object.row.getTable().getName();
+            throw new IllegalArgumentException("Object is of type " + objectClassName + ". Expected " + expectedClass);
+        }
+    }
+
+    private void checkValidIndex(int location) {
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: I would rename this to `isIndexValid()`

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+ */
+public class DynamicRealmObject {
+
+     Realm realm;
+     Row row;
+
+    /**
+     * Creates a dynamic Realm object based on a row entry.
+     */
+    public DynamicRealmObject(Realm realm, Row row) {
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: I'm not sure I like this constructor since it exposes Row. How about having `DynamicRealmObject` extend `RealmObject` or something from which RealmObject could extend from. That was we could have `realm` and `row` in `RealmObject` as `protected` and be able to use them here. So the constructor could simply take a RealmObject.

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+    private Realm realm;
+    private DynamicRealmObject dObj;
+
+    @Override
+    protected void setUp() throws Exception {
+        RealmConfiguration realmConfig = new RealmConfiguration.Builder(getContext()).schema(AllJavaTypes.class).build();
+        Realm.deleteRealm(realmConfig);
+        realm = Realm.getInstance(realmConfig);
+        realm.beginTransaction();
+        AllJavaTypes obj = realm.createObject(AllJavaTypes.class);
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: AllJavaTypes should have it's fields renamed replacing "Column" with "Field" - but in a separate PR.

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+    @Override
+    protected void tearDown() throws Exception {
+        super.tearDown();
+        realm.close();
+    }
+
+    // Test invalid input: <empty, non-existing field, wrong field type>
+    public void testGetBooleanIllegalArguments() {
+        List<String> arguments = Arrays.asList(null, "foo", "columnString");
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: use constants instead of "columnString" etc.

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        for (String argument : arguments) {
+            try {
+                dObj.getBoolean(argument);
+                fail(argument + " should throw exception.");
+            } catch (IllegalArgumentException expected) {
+            }
+        }
+    }
+
+    public void testGetBoolean() {
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: All these tests are so repetitive but dispersed. I would suggest to make a single testcase testing all GetXX, a single test case testing GetLinkedXX() and a third testing GetXXIllegalArguments.
That would be so much easier to get an overview of.

File: realm/src/androidTest/java/io/realm/entities/AllJavaTypes.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+import io.realm.RealmList;
+import io.realm.RealmObject;
+import io.realm.annotations.Ignore;
+import io.realm.annotations.Index;
+import io.realm.annotations.PrimaryKey;
+
+public class AllJavaTypes extends RealmObject{
+
+    @Ignore private String columnIgnored;
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/entities/AllJavaTypes.java]: update naming to "field" instead of "column"

Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/entities/AllJavaTypes.java]: add constants to the names so they can be used allover

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    public int size() {
+        return ((Long)linkView.size()).intValue();
+    }
+
+    private void checkIsValidObject(DynamicRealmObject object) {
+        if (object == null) {
+            throw new IllegalArgumentException("DynamicRealmList does not accept null values");
+        }
+        if (!realm.getPath().equals(object.realm.getPath())) {
+            throw new IllegalArgumentException("Cannot add a object belonging already in another Realm");
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: a -> an

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        if (row.getIndex() != other.row.getIndex()) return false;
+
+        return true;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString(); // TODO How to iterate across all fields?
+    }
+}
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: missing newline

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+     * Creates a dynamic Realm object based on a row entry.
+     */
+    public DynamicRealmObject(Realm realm, Row row) {
+        this.realm = realm;
+        this.row = row;
+    }
+
+    public boolean getBoolean(String fieldName) {
+        long columnIndex = row.getColumnIndex(fieldName);
+        checkFieldExists(columnIndex, fieldName);
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Should the validation be here and not in C++?
Doesn't Row already do the checks?
If it should be here, make one method to validate all instead of 3 method calls.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        int result = 17;
+        result = 31 * result + ((realmName != null) ? realmName.hashCode() : 0);
+        result = 31 * result + ((tableName != null) ? tableName.hashCode() : 0);
+        result = 31 * result + (int) (rowIndex ^ (rowIndex >>> 32));
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: return statements on separate line.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java (hunk 5 above)
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: everywhere else we use "check" for these checks - so keep that or change it everywhere else. "check" is better that "is" due to missing boolean return type. But "validateIndex" may be better than "checkValidIndex", but again requires that it's changed everywhere else. If that should be done it must be a separate PR.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        return true;
+    }
+
+    @Override
+    public void clear() {
+        linkView.clear();
+    }
+
+    @Override
+    public DynamicRealmObject get(int location) {
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: "location" -> "index" (that's also the standard for AbstractList). Update everywhere.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    public DynamicRealmObject set(int location, DynamicRealmObject object) {
+        checkIsValidObject(object);
+        checkValidIndex(location);
+        linkView.set(location, object.row.getIndex());
+        return object;
+    }
+
+    @Override
+    public int size() {
+        return ((Long)linkView.size()).intValue();
```
Comment (bmunkholm) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: How to handle lists larger than int? If we don't we might as well limit everything in Java to only handle int sized Realms.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 6 above)
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: I agree, exposing internal classes is bugging me as well. I considered a few other options but neither seem that great either:

1) Moving DynamicRealmObject/DynamicRealmList to ```io.realm``` instead of ```io.realm.dynamic``` -> I would prefer to keep the main package space as clean as possible, i.e. only primary classes/interfaces.

2) Extending RealmObject doesn't work if we keep these classes in ```.dynamic``` as they cannot access the package protected fields from RealmObject.

Generally the current approach where we rely package protected methods/fields to avoid exposing internals is hurting our ability to modularise the code base. I don't have any golden answer though. But we should probably give this some more thought...

Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: About 2) If you make the fields `protected` (not `package protected`) it should work!

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java (hunk 16 above)
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: What do you mean handle it? Even if every row only used 1 byte, a list with size = max_int would still hold 2 GB of data, so this would not be relevant for a mobile application?    

Fun fact from the Collections JavaDoc: "size(): Returns the number of elements in this collection. If this collection contains more than Integer.MAX_VALUE elements, returns Integer.MAX_VALUE."

So we should actually do a ```Math.min(linkView.size(), Integer.MAX_VALUE)``` here if want to honour that contract.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 13 above)
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: No, row is non-checking as it up until now has been hidden behind the typed API where it couldn't fail. So error checking should be done here avoid adding unnecessary overhead to the typed API. Checks could probably be merged though.

Comment (kneth) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Getting a column index, the C++ function will return `not_found` (or -1) if the name if not a valid column name.

The C++ getters for `Row` do check the type, and an exception will be thrown.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+        if (row.getIndex() != other.row.getIndex()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString(); // TODO How to iterate across all fields?
```
Comment (kneth) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: for(int i = 0; i < row.getColumnCount(); i++) {
        ...
    }

**Comment by kneth** [comment]
Why doesn't `RealmDynamicObject` have setters?

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 17 above)
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Smart ass :) My problem was accessing the data since AFAIK there is no generic row.getObject(index) function, so you would have to check the column type for each index in order to use the proper getX function. It's doable sure, but seemed like a lot of work if we had a smarter function. Maybe using the json() function in core?

**Comment by cmelchior** [comment]
I hadn't gotten around to it yet (WIP), and my driver use case didn't require them. But they are incoming :)

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 13 above)
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Hmm, seems you are right @kneth. I wonder if we should rethink this? Because our typed API should really not need the overhead of checking row/col/type on each access, while the dynamic API would?

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 6 above)
Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Actually extending RealmObject doesn't work at all since you would have to parse the original RealmObject in as a parameter. We can't access the realm/row of that class unless they are in the same package or are public :(

```
public DynamicRealmObject(RealmObject obj) {
  // Boom! Unless RealmObject and DynamicRealmObject are in the 
  // same package or obj.row is public.   
  this.row = obj.row; 
}
```

Users shouldn't need to create these anyways. They should always be created using e.g. something like the below when doing migrations. So the JavaDoc should definitely reflect this when adding the new migration API. 

```
DynamicRealmObject obj = tableSpec.createObject();
```

Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Hmmm... I think we should really strive to avoid any reference to internal classes. More brains are needed! @nhachicha @beeender ? Any suggestion?

**Comment by cmelchior** [comment]
This has left WIP and is ready for review: @realm/java 

Currently the constructor for DynamicRealmObject takes a Row object. This is not ideal, so any feedback on work arounds that doesn't involve us exposing internal API's would be helpful.

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 18]
```diff
@@ -367,7 +367,11 @@ public String getSimpleClassName() {
      * model classes.
      */
     public boolean isModelClass() {
-        return (!classType.toString().endsWith(".RealmObject") && !classType.toString().endsWith("RealmProxy"));
+        String type = classType.toString();
+        if (type.equals("io.realm.dynamic.DynamicRealmObject")) {
+            return false;
+        }
+        return (!type.endsWith(".RealmObject") && !type.endsWith("RealmProxy"));
```
Comment (kneth) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Can't you collapse it to one line?

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        } finally {
+            realm.cancelTransaction();
+            realm2.close();
+        }
+    }
+
+    public void testAddObject() {
+        realm.beginTransaction();
+        AllJavaTypes obj = realm.createObject(AllJavaTypes.class);
+        dynamicList.add(new DynamicRealmObject(obj));
```
Comment (kneth) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: Checking that `dynamicList` only has one element before`add` would ensure that `add` only adds one element.

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+        List<String> arguments = Arrays.asList(null, "foo", AllJavaTypes.FIELD_STRING, linkedField);
+        List<String> stringArguments = Arrays.asList(null, "foo", AllJavaTypes.FIELD_BOOLEAN, linkedField);
+
+        // Test all getters
+        for (SupportedType type : SupportedType.values()) {
+            List<String> args = (type == SupportedType.STRING) ? stringArguments : arguments;
+            try {
+                callGetter(type, args);
+                fail();
+            } catch(IllegalArgumentException expected) {
```
Comment (kneth) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: `ignored`

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+        List<String> arguments = Arrays.asList(null, "foo", AllJavaTypes.FIELD_STRING, linkedField);
+        List<String> stringArguments = Arrays.asList(null, "foo", AllJavaTypes.FIELD_BOOLEAN, linkedField);
+
+        // Test all getters
+        for (SupportedType type : SupportedType.values()) {
+            List<String> args = (type == SupportedType.STRING) ? stringArguments : arguments;
+            try {
+                callSetter(type, args);
+                fail();
+            } catch(IllegalArgumentException expected) {
```
Comment (kneth) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: `ignored`

File: realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+    public void testSetObjectWrongTypeThrows() {
+        realm.beginTransaction();
+        AllJavaTypes obj = realm.createObject(AllJavaTypes.class);
+        Dog otherObj = realm.createObject(Dog.class);
+        DynamicRealmObject dynamicObj = new DynamicRealmObject(obj);
+        DynamicRealmObject dynamicWrongType = new DynamicRealmObject(otherObj);
+        try {
+            dynamicObj.setObject(AllJavaTypes.FIELD_OBJECT, dynamicWrongType);
+            fail();
+        } catch (IllegalArgumentException ignore) {
```
Comment (kneth) [inline realm/src/androidTest/java/io/realm/DynamicRealmObjectTest.java]: `ignored`

File: realm-jni/src/io_realm_internal_UncheckedRow.cpp [hunk 23]
```diff
@@ -288,3 +288,10 @@ JNIEXPORT jboolean JNICALL Java_io_realm_internal_UncheckedRow_nativeIsAttached
     TR_ENTER_PTR(nativeRowPtr)
     return ROW(nativeRowPtr)->is_attached();
 }
+
+JNIEXPORT jboolean JNICALL Java_io_realm_internal_UncheckedRow_nativeHasField
```
Comment (bmunkholm) [inline realm-jni/src/io_realm_internal_UncheckedRow.cpp]: This method seems to have no value. Why not just make this "helper" in Java

File: realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java (hunk 18 above)
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Yes
```return !type.equals("io.realm.dynamic.DynamicRealmObject") && (!type.endsWith(".RealmObject") && !type.endsWith("RealmProxy"))```

Which is horrible to read and understand IMO

File: realm-jni/src/io_realm_internal_UncheckedRow.cpp (hunk 23 above)
Comment (cmelchior) [inline realm-jni/src/io_realm_internal_UncheckedRow.cpp]: Actually it does. In order to consolidate error messages you can no longer use getColumnIndex(fieldName) with illegal field names, it will throw an exception instead of returning -1. So the implicit behaviour of getColumnIndex to check for existence now require an explicit function.

File: realm/src/main/java/io/realm/RealmObject.java [hunk 24]
```diff
@@ -106,4 +106,22 @@ public void removeFromRealm() {
...
+     * @return The Realm this object belongs to or {code null} if it is a standalone object.
+     */
+    protected static Realm getRealm(RealmObject obj) {
+        return obj.realm;
+    }
+
+    /**
+     * Returns the {@link Row} representing this object. Internal use only.
+     *
+     * @return The {@link Row} his object belongs to or {code null} if it is a standalone object.
```
Comment (nhachicha) [inline realm/src/main/java/io/realm/RealmObject.java]: _his object_ ==> this object

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+    public int size() {
+        long size = linkView.size();
+        return size < Integer.MAX_VALUE ? (int) size : Integer.MAX_VALUE;
+    }
+
+    private void checkIsValidObject(DynamicRealmObject object) {
+        if (object == null) {
+            throw new IllegalArgumentException("DynamicRealmList does not accept null values");
+        }
+        if (!realm.getPath().equals(object.realm.getPath())) {
```
Comment (nhachicha) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: shouldn't we check for the configuration instead?

File: realm/src/androidTest/java/io/realm/TestHelper.java [hunk 26]
```diff
@@ -30,6 +30,9 @@
 import java.util.Random;
 import java.nio.charset.Charset;
 
+import io.realm.internal.CheckedRow;
+import io.realm.internal.Row;
+
```
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/TestHelper.java]: unused imports

**Comment by cmelchior** [comment]
Bump @realm/java

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 27]
```diff
@@ -0,0 +1,513 @@
...
+public class DynamicRealmObject extends RealmObject {
+
+     Realm realm;
+     Row row;
+
+    /**
+     * Creates a dynamic Realm object based on a existing object.
+     */
+    public DynamicRealmObject(RealmObject obj) {
+        Row row = RealmObject.getRow(obj);
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: You should check for `obj` being `null` before using it

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 28]
```diff
@@ -0,0 +1,513 @@
...
+     */
+    public void setObject(String fieldName, DynamicRealmObject value) {
+        long columnIndex = row.getColumnIndex(fieldName);
+        if (value == null) {
+            row.nullifyLink(columnIndex);
+        } else {
+            if (value.realm == null || value.row == null) {
+                throw new IllegalArgumentException("Cannot link to objects that are not part of the Realm.");
+            }
+            if (value.realm != null && !realm.getPath().equals(value.realm.getPath())) {
```
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Should configuration be used instead of path here

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+            String expectedClass = linkView.getTable().getName();
+            String objectClassName = object.row.getTable().getName();
+            throw new IllegalArgumentException("Object is of type " + objectClassName + ". Expected " + expectedClass);
+        }
+    }
+
+    private void checkValidIndex(int index) {
+        long size = linkView.size();
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(String.format("Invalid index: %s. Valid range is [%s, %s]", index, 0, size - 1));
```
Comment (kneth) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: The format should be `%d` and not `%s`.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 30]
```diff
@@ -0,0 +1,516 @@
...
+
+    /**
+     * Creates a dynamic Realm object based on a existing object.
+     */
+    public DynamicRealmObject(RealmObject obj) {
+        if (obj == null) {
+            throw new IllegalArgumentException("Non-null object must be provided.");
+        }
+        Row row = RealmObject.getRow(obj);
+        if (obj == null || row == null) {
```
Comment (kneth) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: How can `obj` be `null`?

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 31]
```diff
@@ -0,0 +1,516 @@
...
+     */
+    public void setObject(String fieldName, DynamicRealmObject value) {
+        long columnIndex = row.getColumnIndex(fieldName);
+        if (value == null) {
+            row.nullifyLink(columnIndex);
+        } else {
+            if (value.realm == null || value.row == null) {
+                throw new IllegalArgumentException("Cannot link to objects that are not part of the Realm.");
+            }
+            if (value.realm != null && !realm.getConfiguration().equals(value.realm.getConfiguration())) {
```
Comment (kneth) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: When can `value.realm` be `null`?

File: realm/src/main/java/io/realm/internal/UncheckedRow.java [hunk 32]
```diff
@@ -283,4 +288,5 @@ public boolean isAttached() {
     protected native void nativeNullifyLink(long nativeRowPtr, long columnIndex);
     protected static native void nativeClose(long nativeRowPtr);
     protected native boolean nativeIsAttached(long nativeRowPtr);
+    protected native boolean nativeHasField(long nativeRowPtr, String columnName);
```
Comment (kneth) [inline realm/src/main/java/io/realm/internal/UncheckedRow.java]: Can be name be `nativeHasColumn`? It is consistent with the other column functions for example `nativeGetColumnCount()`.

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 31 above)
Comment (emanuelez) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: when it's a standalone object

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java [hunk 33]
```diff
@@ -0,0 +1,205 @@
...
+        obj.setFieldInt(1);
+        obj.setFieldLong(1);
+        obj.setFieldFloat(1.23f);
+        obj.setFieldDouble(1.234d);
+        obj.setFieldBinary(new byte[]{1, 2, 3});
+        obj.setFieldBoolean(true);
+        obj.setFieldDate(new Date(1000));
+        obj.setFieldObject(obj);
+        obj.getFieldList().add(obj);
+        dynamicObject = new DynamicRealmObject(obj);
```
Comment (beeender) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: Byte is missing in this Test.

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java [hunk 34]
```diff
@@ -0,0 +1,205 @@
...
+        }
+    }
+
+    public void testGet() {
+        DynamicRealmObject listObject = dynamicList.get(0);
+        assertEquals(dynamicObject, listObject);
+    }
+
+    public void testSetIllegalLocationThrows() {
+        int[] indexes = new int[] { -1, 1 };
```
Comment (beeender) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: Don't you have to beginTransaction first?

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java (hunk 31 above)
Comment (kneth) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Just a few lines above, an exception is thrown if `value.realm` is `null`. If we get to this line, `value.realm` cannot be `null`, right?

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java (hunk 34 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: We are testing for index before checking for beginTransaction so the test is correct. Although you could argue that the current test depends on the internal implementation of ```set()```. I can change  it if you think it is necessary?

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java (hunk 33 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: It has been added

File: realm/src/androidTest/java/io/realm/DynamicRealmListTest.java (hunk 34 above)
Comment (beeender) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: Oh, yes, I want to argue that exactly. I still would like to have that to avoid confusion in the future  :smile:

Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/DynamicRealmListTest.java]: I'll change it :)

**Comment by cmelchior** [comment]
All comments should be fixed: @realm/java

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java [hunk 35]
```diff
@@ -0,0 +1,139 @@
...
+     *
+     * @param object the object to add.
+     * @return true
+     * @throws IllegalArgumentException if object is either {@code null} or has the wrong type.
+     */
+    @Override
+    public boolean add(DynamicRealmObject object) {
+        checkIsValidObject(object);
+        linkView.add(object.row.getIndex());
+        return true;
```
Comment (beeender) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: A rare case could happen here. If the `size()` has been reached `Integer.MAX_VALUE`, `add` will succeed and return true. But user can never get the Object. Should it be more clear that just throw a IndexOutOfBoundsException here?

Comment (cmelchior) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: Actually this is the specified behaviour. Acessing objects after Integer.MAX require an iterator. Blame the original Java API designer ☺

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 36]
```diff
@@ -0,0 +1,526 @@
...
+ * Using a DynamicRealmObject is slower than using the regular model class.
+ */
+public class DynamicRealmObject extends RealmObject {
+
+     Realm realm;
+     Row row;
+
+    /**
+     * Creates a dynamic Realm object based on a existing object.
+     */
```
Comment (beeender) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Param doc missed

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 37]
```diff
@@ -0,0 +1,526 @@
...
+            return new DynamicRealmObject(realm, linkRow);
+        }
+    }
+
+    /**
+     * Returns the {@link io.realm.RealmList} of objects being linked to from this field. This list is returned
+     * as a {@link DynamicRealmList}.
+     *
+     * @param fieldName Name of field.
+     * @return The {@link DynamicRealmList} representation of the RealmList.
```
Comment (beeender) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Lower case "The" following `@return`?

File: realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java [hunk 38]
```diff
@@ -0,0 +1,526 @@
...
+        if (fieldName == null || fieldName.isEmpty()) {
+            return false;
+        }
+        return row.hasColumn(fieldName);
+    }
+
+    /**
+     * Returns the list of field names on this object.
+     *
+     * @return List of field names on this objects or the empty list if the object doesn't have any fields.
```
Comment (beeender) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmObject.java]: Same, "List" -> "list"

File: realm/src/main/java/io/realm/dynamic/DynamicRealmList.java (hunk 35 above)
Comment (beeender) [inline realm/src/main/java/io/realm/dynamic/DynamicRealmList.java]: Ha, got it.

**Comment by kneth** [comment]
:+1:

**Comment by nhachicha** [comment]
:+1:

**Comment by emanuelez** [comment]
:+1: :shipit:

**Comment by beeender** [comment]
+1
